### PR TITLE
Improvements  on acquiring locks 

### PR DESCRIPTION
**Pull request by werberson**
- Fix error of locks immediately removed after being acquired
- Wait for all read locks to be released to allow a write lock

### CODE REVIEW

File: etcdlock/rwmutex2.go [hunk 1]
```diff
@@ -31,16 +29,12 @@ func (rwm *RWMutex) RLock(ctx context.Context) error {
...
-		if done, werr := rwm.waitOnLastRev(ctx, rwm.pfx+"write"); done || werr != nil {
-			err = rwm.myKey.Delete()
-			if err != nil {
-				return err
-			}
-			return werr
-		}
+	// wait until nodes with "/write" and a lower revision number than myKey are gone
+	if err = rwm.waitOnLastRev(ctx, rwm.pfx+"write"); err != nil {
+		_ = rwm.myKey.Delete()
```
Comment (flaviostutz) [inline etcdlock/rwmutex2.go]: Why did you ignore the possible Delete error here?

**Review by flaviostutz — COMMENTED** [review]
@werberson, great work you did here! Congrats!
I commented on a line during review. Please check it out.

File: etcdlock/rwmutex2.go (hunk 1 above)
Comment (werberson) [reply etcdlock/rwmutex2.go]: if RLock or RWLock returns the error from Delete function the real reason of "why was not possible to get a lock" will be masked.

A type of error the function waitOnLastRev can return is "context deadline exceeded" as this is an important error message it should not be overwritten.

Maybe concatenating the errors' messages can be a solution. What do you think?

**Comment by flaviostutz** [comment]
Sure. I think we should concatenate the errors or risk hiding a possible bug or failure.
Thanks again!

**Comment by werberson** [comment]
I did a commit concatenating the errors' messages.

**Comment by flaviostutz** [comment]
@werberson thanks again! Great job!

**Pull request merged**